### DISCUSSION TITLE
Example with complex data and nested array

### DISCUSSION THREAD

**Discussion by gaedlld** [LegendApp/legend-state#297] [General]
Hello, do you have any example or good practice about how to use LegendState with array of objects and nested arrays ?
I'm trying to have a global list of tasks, and each task have a list of sub-tasks (and maybe others nested array). And I would like to manipulate them easily like by having functions and computed getters inside object (like with class). But by following the documentation I'm only able to have static functions which every time manipulate the root list and iterate over each nodes to access the values I want to compute...

**Comment by marbemac** [comment]
👍 a more complex example showcasing how to effectively manage a larger data structure that includes 2+ levels of relationships would be very helpful. Particularly re best practices around computed, and how to manage relationships between several entities. A Linear-esque data model is a good example (users have many issues, issue has many comments, comment has one user, etc). It feels as though legend would be really great to model out a reactive object graph like this, but the best practices for doing so are not well captured in the docs imho.

**Comment by mkhennoussi** [comment]
Honestly, after using LegendState on a project with complex data and nested data, I really regret it, as updates comes from you don't know where exactly, the app crashes you don't know why. I am currently give it a last chance by modeling again (for the 3rd time) the state by avoiding nested data and I will see. I now quite understand the need of a unique source of truth for the state 😂

**Reply by marbemac** [reply]
> I now quite understand the need of a unique source of truth for the state

What do you mean - isn't the legendstate store the unique source of truth for your state?

> Honestly, after using LegendState on a project with complex data and nested data, I really regret it, as updates comes from you don't know where exactly, the app crashes you don't know why

This is what is making me nervous as I experiment with legend + more complicated real-world data structures. The examples in the docs are nice and simple, and the concepts sound great, but I'm struggling a bit to find a sweet spot for more complex nested structures.

Out of curiosity, are you able to share what your data structure / shape looks like (or a simplified version of it that still showcases the bits that are tricky) as an example of another more real world example that an example would be great for?

**Reply by jmeistrich** [reply]
How do you think we could do it better? The next area I want to focus on is improving debugging. Unfortunately the batching makes call stacks not work well. Some ideas I have:

1. A way to list all changes in the current batch
2. Possibly some way to save call stacks, but that would impact performance significantly so would only be useful in dev
3. Integrating with dev tools to show the current state of each observable

Can you share some examples of bugs that are hard to debug? Or do you have any ideas for how you'd like to see debugging work better?

**Reply by mkhennoussi** [reply]
Hi guys, after reading my comment again, I think it's a little bit too harsh, sorry for that, I didn't want to. 

Thanks again for your amazing work @jmeistrich !

I suggest, before focusing on having more debugging tools, we share a way of modeling complex data with LegendState and/or having best practices. I had a lot of "non sense" things happening with arrays and computed (I know v3 is in alpha).  

Here is some "non trivial" examples made by jotai : https://jotai.org/docs/recipes/large-objects and the codesandbox that comes with it : https://codesandbox.io/s/zealous-sun-f2qnl?file=/src/App.tsx . 

We should also have a clear example with complex data handled inside AND outside react in the same time (https://jotai.org/docs/guides/using-store-outside-react).

I think by getting these kind of advanced examples in the docs, could help people to not be afraid of using LegendState in production.

**Reply by jmeistrich** [reply]
@mkhennoussi Gotcha, I understand :). There have been a whole lot of requests for more examples, and I think we should definitely focus on that as v3 is getting more stable.

We have this page which was intended to show some examples of different ways to use it, but in retrospect maybe it's poorly named, and it's very barebones! https://legendapp.com/open-source/state/v3/guides/patterns/

Can you share more about what nonsense things happened and what you're struggling with structuring? Hopefully I can help you directly and that would help to know what needs to be documented better.
